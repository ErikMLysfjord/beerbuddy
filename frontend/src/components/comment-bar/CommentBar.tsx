import { Button, Input, Spin, App } from "antd";
import styles from "./CommentBar.module.css";
import { useId, useState } from "react";
import { useParams } from "react-router-dom";
import protectRoute from "../../utils/protectRoute";
import useWindowDimensions from "../../utils/useWindowDimensions";

interface CommentBarInterface {
  onSuccess: () => void;
}

/**
 * Posts a comment to the backend.
 * @param beerId - The ID of the beer to post a comment for.
 * @param comment - The comment to post.
 * @returns - The response from the backend.
 */
const postComment = async (beerId: string, comment: string) => {
  if (await protectRoute()) return "Error";
  const userId = localStorage.getItem("userIdBeerBuddy");
  const query = {
    query: `{ comment(userId: "${userId}", beerId: ${beerId}, comment: "${comment}") }`,
  };

  return await fetch(import.meta.env.VITE_APP_BACKEND_URL + "/action", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Accept: "application/json",
    },
    body: JSON.stringify(query),
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error("Network response was not ok");
      }
      return response.json();
    })
    .catch((error) => {
      console.error("There was a problem with the fetch operation:", error);
      return "Error";
    });
};

/**
 * The comment bar component. Contains the input field and submit button for posting a comment.
 * @param onSuccess - The function to call when a comment has been successfully posted.
 * @returns - The comment bar component.
 */
const CommentBar = ({ onSuccess }: CommentBarInterface) => {
  const { id } = useParams<{ id: string }>();
  const [commentText, setCommentText] = useState("");
  const { message } = App.useApp();
  const { width } = useWindowDimensions();
  const inputId = useId();

  if (id === undefined)
    return (
      <div style={{ display: "flex", justifyContent: "center" }}>
        <Spin size="default" />
      </div>
    );

  /**
   * Checks if the comment is valid.
   * A valid comment is a comment that is between 1 and 200 characters long
   * A valid comment does not start with a whitespace.
   * A valid comment cant only contain special characters.
   * @param comment
   * @returns false if the comment is invalid, true if the comment is valid.
   */
  const validComment = (comment: string) => {
    const commentRegex = /^(?! )[^\n].{0,199}$/;
    const onlySpecialCharsRegex = /^[^a-zA-Z0-9]+$/;

    if (!(commentRegex.test(comment) && !onlySpecialCharsRegex.test(comment))) {
      message.error("Your comment is invalid.");
      return false;
    }
    return true;
  };

  /**
   * Handles the posting of a comment.
   * if a comment is invalid, a message will be posted to user and the function returns.
   * if the response is "Error", a message is posted to the user.
   * If the response is successful, the comment text is cleared and onSuccess is called.
   */
  const handleComment = async () => {
    if (!validComment(commentText)) {
      return;
    }
    const response = await postComment(id, commentText);
    if (response === "Error") {
      message.error("There was a problem posting your comment.");
      return;
    }
    message.success("Comment posted.");
    setCommentText("");
    onSuccess();
  };

  return (
    <div className={styles.wrapper}>
      <div className={styles.divider} />
      <section className={styles.container} aria-label="Post a comment">
        <div className={styles.labelContainer}>
          <label htmlFor={inputId}>Comment</label>
          <Input
            id={inputId}
            placeholder="Best beer ever!"
            className={styles.inputField}
            value={commentText}
            onChange={(e) => setCommentText(e.target.value)}
            onKeyDown={(e) => {
              /* When you press Enter with focus on input, you post */
              if (e.key === "Enter") {
                handleComment();
              }
            }}
          />
        </div>
        {width > 768 ? (
          <Button
            type="primary"
            className={styles.submitButton}
            /* When clicking on post button, you post */
            onClick={handleComment}
          >
            Comment
          </Button>
        ) : (
          <Button type="primary" className={styles.submitButton}>
            <img
              width={"30px"}
              height={"30px"}
              alt="Send icon"
              src={"/project2/paper-plane-right.svg"}
            />
          </Button>
        )}
      </section>
    </div>
  );
};
export default CommentBar;
