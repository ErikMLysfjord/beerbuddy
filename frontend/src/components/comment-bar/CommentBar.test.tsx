import { Mock, vi } from "vitest";
import { render, fireEvent, waitFor, act } from "@testing-library/react";
import CommentBar from "./CommentBar";
import { axe } from "jest-axe";

const mockError = vi.fn();
const mockSuccess = vi.fn();

/**
 * Mock the App context to return a mock message object, with success and error functions.
 */
vi.mock("antd", async () => {
  const actual: object = await vi.importActual("antd");
  return {
    ...actual,
    App: {
      useApp: () => ({ message: { error: mockError, success: mockSuccess } }),
    },
  };
});

/**
 * Mock useParams to return a mock ID.
 */
vi.mock("react-router-dom", () => ({
  useParams: () => ({ id: "1" }),
}));

/**
 * Mock fetch to always return a successful response.
 */
global.fetch = vi.fn(() =>
  Promise.resolve({
    ok: true,
    json: () => Promise.resolve(),
  })
) as Mock;

describe("CommentBar", () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it("should match snapshot", () => {
    const { container } = render(<CommentBar onSuccess={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  it("should be accessible", async () => {
    const { container } = render(<CommentBar onSuccess={() => {}} />);
    expect(await axe(container)).toHaveNoViolations();
  });

  it("should render the input field and submit button", () => {
    const { getByPlaceholderText, getByRole } = render(
      <CommentBar onSuccess={() => {}} />
    );
    expect(getByPlaceholderText("Best beer ever!")).toBeInTheDocument();
    expect(getByRole("button", { name: "Comment" })).toBeInTheDocument();
  });

  it("should call onSuccess when a comment is successfully posted", async () => {
    const onSuccess = vi.fn();
    const { getByPlaceholderText, getAllByText } = render(
      <CommentBar onSuccess={onSuccess} />
    );
    const input = getByPlaceholderText("Best beer ever!");
    const button = getAllByText("Comment")[1];

    fireEvent.change(input, { target: { value: "This is a comment" } });
    fireEvent.click(button);

    await waitFor(() => expect(onSuccess).toHaveBeenCalled());
  });

  it("should post a message when pressing enter on input field", async () => {
    const { getByPlaceholderText } = render(
      <CommentBar onSuccess={() => {}} />
    );
    const input = getByPlaceholderText("Best beer ever!");

    fireEvent.change(input, { target: { value: "This is a comment" } });
    fireEvent.keyDown(input, { key: "Enter", code: "Enter" });

    await waitFor(() => expect(mockSuccess).toHaveBeenCalled());
  });

  it("should display a success message when posting a comment succeeds", async () => {
    const { getByPlaceholderText, getAllByText } = render(
      <CommentBar onSuccess={() => {}} />
    );
    const input = getByPlaceholderText("Best beer ever!");
    const button = getAllByText("Comment")[1];

    fireEvent.change(input, { target: { value: "This is a comment" } });
    fireEvent.click(button);

    await waitFor(() => expect(mockSuccess).toHaveBeenCalled());
  });

  it("should display an error message when posting a comment fails", async () => {
    /* Mock fetch to always return an unsuccessful response */
    global.fetch = vi.fn(() =>
      Promise.resolve({
        ok: false,
        json: () => Promise.resolve(),
      })
    ) as Mock;

    const { getByPlaceholderText, getAllByText } = render(
      <CommentBar onSuccess={() => {}} />
    );
    const input = getByPlaceholderText("Best beer ever!");
    const button = getAllByText("Comment")[1];

    fireEvent.change(input, { target: { value: "This is a comment" } });
    fireEvent.click(button);

    await waitFor(() => expect(mockError).toHaveBeenCalled());
  });

  it("Should Throw error on empty comment", async () => {
    const { getByPlaceholderText, getAllByText, queryByRole, queryAllByText } =
      render(<CommentBar onSuccess={() => {}} />);
    const input = getByPlaceholderText("Best beer ever!");
    const button = getAllByText("Comment")[1];

    fireEvent.change(input, { target: { value: "    " } });
    fireEvent.click(button);
    expect(queryByRole("img")).not.toBeInTheDocument();
    expect(queryAllByText("Comment")).toHaveLength(2);

    await waitFor(() => expect(mockSuccess).not.toHaveBeenCalled());
  });

  it("should throw error for a comment with more than 280 characters", async () => {
    const { getByPlaceholderText, getAllByText } = render(
      <CommentBar onSuccess={() => {}} />
    );
    const input = getByPlaceholderText("Best beer ever!");
    const button = getAllByText("Comment")[1];

    fireEvent.change(input, { target: { value: "a".repeat(281) } });
    fireEvent.click(button);

    await waitFor(() => expect(mockSuccess).not.toHaveBeenCalled());
    await waitFor(() => expect(mockError).toHaveBeenCalled());
  });

  it("should throw error for a comment that only contains special characters", async () => {
    const { getByPlaceholderText, getAllByText } = render(
      <CommentBar onSuccess={() => {}} />
    );
    const input = getByPlaceholderText("Best beer ever!");
    const button = getAllByText("Comment")[1];

    fireEvent.change(input, {
      target: { value: "!!!@#$%^&*()_+{}[]|;'<>,.?/~`" },
    });
    fireEvent.click(button);

    await waitFor(() => expect(mockSuccess).not.toHaveBeenCalled());

    it("should render a button without text when the screen width is less than 768px", () => {
      const { getByAltText, queryByText, queryByRole } = render(
        <CommentBar onSuccess={() => {}} />
      );

      expect(queryByRole("img")).not.toBeInTheDocument();
      expect(queryByText("Comment")).toBeInTheDocument();

      // Mock the window width to be 500px.
      // This is done to ensure the correct text is rendered.
      global.innerWidth = 500;
      act(() => {
        global.dispatchEvent(new Event("resize"));
      });

      // The button should now have a send icon instead of "Comment".
      expect(getByAltText("Send icon")).toBeInTheDocument();
      expect(queryByText("Comment")).not.toBeInTheDocument();
      expect(queryByRole("img")).toBeInTheDocument();
    });
  });
});
