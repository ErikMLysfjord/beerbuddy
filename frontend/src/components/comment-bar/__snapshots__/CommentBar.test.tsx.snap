// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CommentBar > should match snapshot 1`] = `
<div>
  <div
    class="_wrapper_18a6db"
  >
    <div
      class="_divider_18a6db"
    />
    <section
      class="_container_18a6db"
    >
      <div
        class="_labelContainer_18a6db"
      >
        <label
          for=":r0:"
        >
          Comment
        </label>
        <input
          class="ant-input css-dev-only-do-not-override-pr0fja _inputField_18a6db"
          id=":r0:"
          placeholder="Best beer ever!"
          type="text"
          value=""
        />
      </div>
      <button
        class="ant-btn css-dev-only-do-not-override-pr0fja ant-btn-primary _submitButton_18a6db"
        type="button"
      >
        <span>
          Comment
        </span>
      </button>
    </section>
  </div>
</div>
`;
